### PR TITLE
Utility for increasing budget during execution of PyTeal applications

### PR DESCRIPTION
**Pull request by algoidurovic**
This PR implements a PyTeal utility for increasing the opcode budget during execution of an app call using inner app call transactions.

Two modes are provided:
1. Explicit: the app to call for increasing budget is provided by the user.
2. OnCall: the app to call is created during each request to increase the budget, then deleted afterwards.

Two budget increase methods are provided:
1. ensure_budget(budget): increase the current opcode budget to at least the budget requested by the user.
2. maximize_budget(fee): increase the current opcode budget as much as possible without spending more than the given fee. This can be used when the contract developer doesn't want to spend more than the fee provided to increase the budget without needing to do in depth opcode cost analysis on their program. If the fee isn't sufficient for executing the program then a failure may be desired. It can also allow the app caller to specify the maximum fee they want to spend for increasing opcode budget. This behavior can fairly easily be extended to support a global fee limit for budget increase requests. (Inspired by the algofi implementation of opup).

Testing
Unit tests for type checking enforcement and sandbox tests for e2e functionality. The PyTeal programs I used for testing in sandbox are located in examples/application/opup.py

### CODE REVIEW

File: pyteal/ast/opup.py
Comment (barnjamin) [inline pyteal/ast/opup.py:134]: Since we do this check I think I'd favor a name like `ensure`

File: pyteal/ast/opup.py
Comment (jasonpaulos) [inline pyteal/ast/opup.py:33]: If this is used as the app approval program for `OnCreate`, then _anyone_ could update or delete this app, since it does not check who the sender is.

Secondary concern, why is `byte` part of the program string?

File: pyteal/ast/opup.py
Comment (jasonpaulos) [inline pyteal/ast/opup.py:138]: Are you aware that you can create and delete an app in a single transaction? During the app create call, if you set `TxnField.on_completion` to `OnComplete.DeleteApplication` and it will delete the app immediately after it's created.

File: pyteal/ast/opup.py
Comment (jasonpaulos) [inline pyteal/ast/opup.py:134]: Have you considered what happens around boundary conditions? There are two cases I can think of:
* The program already has extremely low budget (e.g. 1 unit left) before entering this function. Likely it will fail because you can't even check if you need to op up successfully with such a small budget, so you should put in the docs what the minimum op budget should be before calling the function.
* The program has only slightly more than the amount it asked for. There _might_ be a situation where this function thinks it has enough budget, but the act of returning from the function causes the budget to drop a few units below the amount the caller asked for (due to branch statements that still need to get executed). I've not verified this is the case, but if you know X additional budget is needed to return from the function, you could check `If(Global.opcode_budget() < required_budget + Int(X))`.

Comment (jasonpaulos) [reply pyteal/ast/opup.py:134]: Another way to phrase the above questions is: how much budget is used before the first inner txn to op up, and how much budget is used after the last `Global.opcode_budget()` check?

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:33]: This was more of a placeholder than anything so haven't put much thought into it. Point taken though.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:134]: Totally agree, `ensure` or `ensure_budget` would be more clear.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:138]: I'm trying to minimize the number of app creations by limiting `execute` to create only one app per invocation, even if reaching the given budget requires multiple inner app calls.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:134]: Good point. A call to `execute` should guarantee that the opcode budget is at least the amount passed by the user **after** the call is completed.

Comment (algoidurovic) [reply pyteal/ast/opup.py]: Fixed. Resolving.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: fixed. resolving.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: is this okay to resolve?

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: fixed. resolving.

**Comment by tzaffi** [comment]
> Two budget increase methods are provided:
> 
> 1. ensure_budget(budget): increase the current opcode budget to at least the budget requested by the user.
> 2. maximize_budget(fee): increase the current opcode budget as much as possible without spending more than the given fee.

I think method 1 is essential, but I'm not as sure about method 2. On the face of it, `maximize_budget()` gives you an unknown amount of extra budget, so you could run out unexpectedly or get much more than you really need. One could argue that method 1 suffers from the opposite problem - what if fees increase and the app can no longer afford to op-up? But that seems a little easier to ameliorate dynamically, if another method is added `fee_for_budget(budget)` which would let the user know how much they'll use up and adapt accordingly.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:127]: @algoidurovic Gauging my understanding - Is the following statement accurate?

If `_ensure_budget_expr` fails, then the created app lives indefinitely.

_If_ my understanding is accurate, I'm wondering if it's possible to avoid leaking the app by creating an inner transaction group via `InnerTxnBuilder` containing (1) create app, (2) op up, and (3) delete app?

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:131]: @algoidurovic For my understanding:  With `OpUpMode.Explicit`, what operations require a buffered budget?

Here's what I interpreted - let me know if/where I'm missing intuition:
*  `OpUpMode.Explicit` does _not_ require a buffered budget because control returns to the application _immediately_ after increasing opcode budget.
* `OpUpMode.OnCall` _does_ require a buffered budget because a delete app operation follows the opcode budget increase.
  * Aside:  How is `self.buffer = Int(50)` determined?
    * If I'm following, it seems the buffer can be specialized for the delete operation.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:136]: @algoidurovic Is it accurate that `maximize_budget` presumes a 1 micro Algo transaction fee?

* I have limited experience here though I know from https://developer.algorand.org/docs/get-details/transactions/#fees that transaction fees are variable.
* I did a brief scan and I don't see a way to gauge the transaction fee without attempting the transaction.
* If I'm following along, then I'm wondering if there's a specific use case that prompted the _maximize_budget_ approach or if it's been added for convenience?
  * I see the intended convenience.  _If_ it's intended as a convenience and _if_ there's there's no reasonable way to account for the dynamic transaction fees, I'm thinking the functionality can be added in a later iteration.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:117]: @algoidurovic and reviewers:  How essential is it to consider apps invoking `ensure_budget` > 1 per invocation?

I raise the use case because I don't know if it's been considered.  And from what I can imagine, the use case _may_ bring substantive PR changes.

* Concerns with multiple invocations:
  * It _may_ be worthwhile to define op up functionality as `@Subroutine`s to minimize program size.  In the PR's current form, the `self` reference prevents `@Subroutine` usage.  Though it's _possible_ to imagine a `@Subroutine` compliant revision. 
  * I can _imagine_ it'd be preferable for `OpUpMode.OnCall` to return the created app ID.  Retaining the app ID reduces opcode cost + transaction fees for subsequent op up calls.  I can imagine a few ways to support the use case though they'll all involve changing the API.
* Provided the use case I'm raising isn't deemed _essential_, I prefer to resolve without changes and retain a simpler approach for now.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:76]: @algoidurovic I'm not sure I follow how the statement is having the intended effect.  Can you help me understand what's happening here?

I _think_ the intent is to updated `target_app_id` for later usage in delete.  However, at the highlighted line, I don't see how the created app ID can be known.  It seems app creation happens _next_.

I'd need to inspect TEAL to be more confident, but I also _believe_ the `load` invocation needs to be part of the returned AST.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:117]: Ah, I realize I missed `self.target_app_id_slot.store(Gitxn[0].created_application_id()),`.  

It seems an attempt is made to cache the created app ID.  However, unless I missed it, I don't see an attempt to read the cached ID.

I'm open to a caching approach.  Though I feel the mode name (or at least docs) should reflect the caching behavior because the semantics are non-obvious and may not match the caller's intent.

**Comment by algoidurovic** [comment]
> > Two budget increase methods are provided:
> > 
> > 1. ensure_budget(budget): increase the current opcode budget to at least the budget requested by the user.
> > 2. maximize_budget(fee): increase the current opcode budget as much as possible without spending more than the given fee.
> 
> I think method 1 is essential, but I'm not as sure about method 2. On the face of it, `maximize_budget()` gives you an unknown amount of extra budget, so you could run out unexpectedly or get much more than you really need. One could argue that method 1 suffers from the opposite problem - what if fees increase and the app can no longer afford to op-up? But that seems a little easier to ameliorate dynamically, if another method is added `fee_for_budget(budget)` which would let the user know how much they'll use up and adapt accordingly.

I updated the PR description to further elaborate on the use cases for maximize_budget(). I do think it's useful to have a fee_for_budget() method available as well.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:127]: Had a verbal with @algoidurovic - I misunderstood inner transaction semantics.  Any failing inner transaction (regardless of grouping) fails the app invocation and all associated transactions. 

So - The PR does _not_ dangle app IDs.  Sorry for the confusion.  Resolving.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:131]: Summarizing a verbal with @algoidurovic:
* I missed the following in my original note:  Since the budget is checked in a while loop, there's _always_ a check after op-ing up to see if the loop completed.  Consequently, there's _always_ a budget needed. 
* Igor to investigate:  Discussion highlighted that it'd probably make sense to invert the comparison ordering.  Concretely, switch from `Global.opcode_budget() < buffered_budget` to `buffered_budget >= Global.opcode_budget()`.  The change ensures `opcode_budget()` is interpreted as late as possible and provides the most up-to-date value.
* Igor to consider:  We discussed that the 50 opcode buffer is intended to be a rough, best fit for many use cases.  Since it's likely too large in some use cases, we discussed optionally offering a configurable override (e.g. factory method) for the app developer to supply a specific value.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:136]: Summarizing a verbal with @algoidurovic:
* There's 2 levels of configurable fees:
  * Congestion fees
  * Min transaction fees configured via consensus parameter
* Igor to investigate:
  * Replace `Int(1000)` with a min transaction fee field to catch consensus parameter changes.
  * Identify if congestion fees apply to inner transactions.  If yes, we ought to highlight that the method _can_ fail at evaluation time by either (or both) changing method name and clarifying docs.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:76]: Summarizing a verbal with @algoidurovic:

* I misinterpreted the AST evaluation.  I mistakenly treated `self.target_app_id` as a local variable.
* We discussed that it's worthwhile to consider different representations that might be less error prone to follow.  Leaving the thread open while the topic is considered though we're good to keep as is.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:117]: Summarizing a verbal with @algoidurovic:
* Igor to consider:  Reworking to support defining as `@Subroutine`.  If that doesn't seem like a good fit, Igor suggested we can point out to the app developer how to wrap invocations in docs.
* We agreed returning the created app ID is out of scope.  It'd be less error prone for the app developer to create the app ID out-of-band and use `Explicit` mode.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:137]: Thanks to @algoidurovic for pointing out that inner transactions are _not_ subject to congestion fees because they're _not_ transmitted over the network.

https://github.com/algorand/pyteal/pull/274/commits/7152af649c6ab0d5a1a6690fcaee95268240b3ea brings in the feedback to factor out the min txn fee constant.    

Resolving.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [inline pyteal/ast/opup.py:105]: @algoidurovic Minor:  Is the intend inequality `<` or `<=`?

I raise the question because I thought it'd be `<=`.  When `=`, I thought there's no need to acquire more opcode budget because the compared budget reflects the budget.

If there's a change to be made, note that there's 2 places to make the change.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py:132]: @algoidurovic Thanks for bringing in some feedback via https://github.com/algorand/pyteal/pull/274/commits/7152af649c6ab0d5a1a6690fcaee95268240b3ea.  

The remaining open question I see is:

> Igor to consider: We discussed that the 50 opcode buffer is intended to be a rough, best fit for many use cases. Since it's likely too large in some use cases, we discussed optionally offering a configurable override (e.g. factory method) for the app developer to supply a specific value.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:105]: > When =, I thought there's no need to acquire more opcode budget because the compared budget reflects the budget.

If I'm understanding you correctly, the existing comparison does enforce this. `<=` would continue the loop when the values being compared are equal.

Comment (michaeldiamant) [reply pyteal/ast/opup.py:105]: @algoidurovic Duh - Sorry for the noise here.  Disregard - resolving.

File: pyteal/ast/opup.py
Comment (jasonpaulos) [inline pyteal/ast/opup.py:118]: I still don't see any docs here saying what the minimum budget should be before calling this function.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: I think we should have a team discussion on coding practices in PyTeal relating to this topic. There's definitely an argument for representing PyTeal "subprograms" as functions rather than variables but this may complicate the code in certain places so we likely shouldn't apply it universally even if we do adopt this practice.

for example:
```
expr = Int(1) + Int(2), program = Seq(..., expr, ...)
->
def expr():
  return Int(1) + Int(2)
program = Seq(..., expr(), ...)
```

The benefit to the second approach is that it's more clear to the developer that `expr` stores an unevaluated expression. With that said, my opinion is that this is somewhat outside the scope of this PR and shouldn't hold up approval. Let me know what you think.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: To your first point, my position is that wrapping is a better option because the overhead is low (in terms of both implementation time/complexity and efficiency) but doesn't prevent the user from using the standard approach, which has less slot usage and fewer arguments. I'll add to the docs as soon as the code is no longer in flux. Open to feedback here.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: I added the option for the user to pass a custom buffer value to the OpUp constructor and tightened the default buffer value after some experimentation. Let me know if this addresses your concerns.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: Should be taken care of now.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py]: The only reason for my original approach was to go easy on indexer but with that concern addressed I agree this approach is better for multiple reasons, one being that it allows for much cleaner code. Made the changes and tested in sandbox.

File: pyteal/ast/opup.py
Comment (jasonpaulos) [inline pyteal/ast/opup.py:32]:
```suggestion
ON_CALL_APP = Bytes("base16", "068101")  # v6 teal program "int 1"
```

Pyteal automatically adds a `return` op at the end of the program, but you can exclude it to save space and 1 extra compute unit (though it's kinda strange that the `return` op consumes budget in the first place 🤔 )

**Review by jasonpaulos — COMMENTED** [review]
Except for my one minor comment, this looks good to me, thank you for incorporating my feedback.

The one comment is about shrinking the on call program code. I edited the TEAL by hand, and even though it's a simple change I'd still recommend you briefly test it in sandbox again just to make sure there are no unintended consequences.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py]: @algoidurovic A few thoughts here below.  I'll resolve because I think we're good though feel welcomed to re-open if you'd like to continue discussion.

A few thoughts:
* Thanks for giving the topic consideration.
* At least for the convention in-question, I don't feel strongly that a single convention _must_ always be followed.  I can imagine different styles making sense in different places.  Which is also to say - I think it's reasonable to try something in 1 place without applying it to _all_ places.
* I agree we need not hold up the PR.  As of https://github.com/algorand/pyteal/pull/274/commits/399a1b915f77dae95eaee41f5837e31c6d5d9ee3, the PR is in a different form and I think the concern is addressed.
* I'm up for a group discussion.  Can make for a post-stand up discussion for those interested in the topic.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py]: @algoidurovic Got it - I haven't considered the topic further.  Let's see how the docs look.  Leaving the thread open as a visual reminder.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py]: @algoidurovic Thanks for bringing in the feedback.  

Before resolving, I have an _optional_ thought.  If you prefer as is, feel welcome to tell me and resolve the thread.

* Idea:  Since `budget` is _only_ referenced by `ensure_budget`, I'd optionally consider moving `buffer` from constructor parameter to method parameter.
* Rationale:  
  * Explicitly demarcates where `buffer` applies.  When provided to the constructor, the user must be aware of the implementation to see usage (e.g. Does `maximize_budget` use `buffer`?). 
  * Additionally, a user may wish to qualify `buffer` on a per invocation basis.  With a constructor parameter, the user must construct separate objects to express the use case.

File: docs/opup.rst
Comment (michaeldiamant) [inline docs/opup.rst:31]: @algoidurovic Minor:  _Suggests_ adding commas for readability.

```suggestion
:any:`Explicit` has the benefit of constructing more lightweight inner transactions, but requires the
target app ID to be provided in the foreign apps array field of the transaction and the :any:`pyteal.OpUp`
constructor in order for it to be accessible. :any:`OnCall` is easier to use, but has slightly more overhead
because the target app must be created and deleted during the evaluation of an app call.
```

File: docs/opup.rst
Comment (michaeldiamant) [inline docs/opup.rst:40]: @algoidurovic Minor thoughts:
* It's possible to link the `Ed25519Verify` reference.   https://stackoverflow.com/questions/22700606/how-would-i-cross-reference-a-function-generated-by-autodoc-in-sphinx helped me understand how to do it.
* Optionally, I prefer to add commas for readability on numbers with > 3 digits.


```suggestion
In the example below, the :py:meth:`pyteal.Ed25519Verify` expression is used, which costs 1,900.
```

File: docs/opup.rst
Comment (michaeldiamant) [inline docs/opup.rst:65]: @algoidurovic Optional _suggestions_:
* Add commas for numbers with > 3 digits.
* Specify units to avoid ambiguity - microAlgos in this case.

```suggestion
In the example below, the fee is capped at 3,000 microAlgos for increasing the budget. This works out to 3 inner
```

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py]: Resolving - As of https://github.com/algorand/pyteal/pull/274/commits/07bd6d25338c3f6da89b38dd9bfcebaceb777e34, `buffer` is no longer needed as a user-configurable parameter.

File: pyteal/ast/opup.py
Comment (michaeldiamant) [reply pyteal/ast/opup.py]: As of https://github.com/algorand/pyteal/pull/274/commits/07bd6d25338c3f6da89b38dd9bfcebaceb777e34, the docs briefly mention wrapping the logic to define as `@Subroutine`. 

I'm _mostly_ comfortable as is.  If program size becomes an issue due to repeated opup calls per app invocation, then we can revisit.  Resolving.

File: docs/opup.rst
Comment (michaeldiamant) [inline docs/opup.rst:3]: @algoidurovic Optionally, I propose mentioning _OpUp_ in the page title.  Since it seems we'll use the language _OpUp_, I think it'd be helpful for readers to see the term in the table of contents + at top of page.

```suggestion
OpUp:  Budget Increase Utility
```

File: examples/application/program.teal
Comment (michaeldiamant) [inline examples/application/program.teal:1]: @algoidurovic Minor:  I'm unfamiliar with our `examples/application` conventions.  I don't see other `.teal` sources versioned.  

I'm wondering if 1 of the following ought to happen?
* Declare we do _not_ version `teal` sources in `examples/application` by adding a `.gitignore` entry.
* Declare we _want_ to version `teal` sources and either rename to a more specific name (e.g. `opup.teal`) or reorganize applications (e.g. dir per app).
* Delete `program.teal` and deal with the concern another time.

If a policy doesn't exist, I think it's useful to eventually declare a policy.  Though I don't want to hold the PR.

I can be convinced to leave the file as is rather than deleting or renaming it.  Though leaving as is feels crufty to me.

**Review by michaeldiamant — APPROVED** [review]
@algoidurovic Thanks for your effort here - I'm excited to see the feature used by the community!  ☕ 

I left an optional remark in https://github.com/algorand/pyteal/pull/274#discussion_r856485694.  I'd appreciate considering it before merging though it need _not_ hold us up.

File: docs/opup.rst
Comment (jasonpaulos) [inline docs/opup.rst:44]: nit: something more explicit saying this should be replaced with an app ID the user provides would help

File: docs/opup.rst
Comment (jasonpaulos) [inline docs/opup.rst:52]: nit: I think `Assert` would be a better example than `Pop`

File: docs/opup.rst
Comment (jasonpaulos) [inline docs/opup.rst:58]: nit: would be nice if this section for `maximize_budget` and the previous section for `ensure_budget` were under different subheadings to break up the explanations & examples

File: docs/opup.rst
Comment (jasonpaulos) [inline docs/opup.rst:66]: I'd say ~700, since there is some overhead

File: docs/opup.rst
Comment (jasonpaulos) [inline docs/opup.rst:13]: nit: it would be nice if the overview explicitly stated that how this works (by executing noop inner txns that increase the group's pooled compute budget) and where the funding comes from (the app's account, not the caller or creator's). The first is a nice-to-have, but the second seems more important for practical purposes

**Review by jasonpaulos — COMMENTED** [review]
The docs look great, I just have some suggestions, only one of which is important (the 700 vs ~700 comment).

File: examples/application/program.teal
Comment (algoidurovic) [reply examples/application/program.teal]: Sorry, this file should've been deleted. Taken care of now.

File: examples/application/opup.py
Comment (jasonpaulos) [inline examples/application/opup.py:5]: This line shouldn't be necessary

**Review by jasonpaulos — COMMENTED** [review]
Looks great, I just have one comment that I missed before

File: pyteal/ast/opup.py
Comment (ahangsu) [inline pyteal/ast/opup.py:33]: minor proposal, do you think `MIN_TXN_FEE` should be available in a larger scope, say it lives in `pyteal/config.py`? I guess some folks would want this available not just in `opup`.

File: pyteal/ast/opup.py
Comment (ahangsu) [inline pyteal/ast/opup.py:132]: just one question: is these lines of code ensuring that in the case of: *current budget slightly higher than required budget*, the `_construct_itxn` will be executed at least once?

Since I am not sure if I understand
> it's possible for ensure_budget() to return with a current budget less than the required budget.

**Review by ahangsu — COMMENTED** [review]
just one question about `ensure_budget`

File: pyteal/ast/opup.py
Comment (jasonpaulos) [reply pyteal/ast/opup.py:33]: `Global.min_txn_fee()` is available to PyTeal, perhaps this code should use that instead?

Comment (ahangsu) [reply pyteal/ast/opup.py:33]: ^ sounds pretty reasonable to me

File: examples/application/program.teal
Comment (michaeldiamant) [reply examples/application/program.teal]: @algoidurovic No worries - Thanks for the change in https://github.com/algorand/pyteal/pull/274/commits/a9d22ea6a16fcf4a0acbab00760395668d94b325.  Resolving.

File: pyteal/ast/opup.py
Comment (algoidurovic) [reply pyteal/ast/opup.py:132]: Say the user calls the method with `required_budget=1000` then, after issuing an inner app call, the current budget is 1000. When the comparison `current_budget < required_budget` is executed it will return false, but the comparison itself consumes budget so the resulting budget may be something like 999 instead of 1000.

**Pull request merged**
